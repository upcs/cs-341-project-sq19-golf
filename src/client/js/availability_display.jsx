import React, { Component } from 'react';
import {Link} from 'react-router-dom';
import ReactTable from "react-table";
import 'react-table/react-table.css'
import '../css/styles.css';

//Availability Table
export class AvailabilityContainer extends Component {
	constructor(props) {
		super(props);

		this.state = {
      showMenu: this.props.showMenu || false,
			dropdownMenu: this.props.dropdownMenu || null,
	  	selected: -1
		};

	  this.showMenu = this.showMenu.bind(this);
		this.closeMenu = this.closeMenu.bind(this);
	}

	showMenu(event) {
		event.preventDefault();

		this.setState({ showMenu: true }, () => {
		  document.addEventListener('click', this.closeMenu);
		});
  }

  closeMenu(event) {
		if (!this.state.dropdownMenu.contains(event.target)) {

		  this.setState({ showMenu: false }, () => {
				document.removeEventListener('click', this.closeMenu);
		  });

		}
  }

	/*handleSubmit(event) {
			event.preventDefault();
			const data = new FormData(event.target);

			fetch('/api/scheduleRequest', {
			  method: 'POST',
			  body: data
			});
  }*/

	render() {
		let data = [];
		for(let time = 8; time <= 22; time++) {
			let row = {
				times: time + ":00",
				monday: null,
				tuesday: null,
				wednesday: null,
				thursday: null,
				friday: null,
				saturday: null,
				sunday: null
			}

			data.push(row);
		}

		let columns = [], colAccessors = ['times', 'monday', 'tuesday', 'wednesday', 'thursday', 'friday', 'saturday', 'sunday'];
		for (let i = 0; i < colAccessors.length; i++) {
			let accessor = colAccessors[i];
			let col = {
				Header: accessor.charAt(0).toUpperCase() + accessor.slice(1),
				accessor: accessor,
				getProps: {(state, rowInfo) => ({
					if (typeof rowInfo !== "undefined") {
						return {
								onClick: (e, handleOriginal) => {
									this.setState({
										selected: rowInfo.row.accessor = 1;
									});
									var timePref = [][];
									function addPref (rowInfo, rowInfo.row.accessor) {
										timePref[rowInfo.row.acessor].push(rowInfo);
									}
									if(handleOriginal) {
										handleOriginal()
									}
								},
								style: {
									background: rowInfo.row.accessor === this.row.acessor? '#730ac9' : 'white',
									color: rowInfo.row === this.row.accessor? 'white' : 'black'
								},
							}
					}
					else {
						return {
							onClick: (e, handleOriginal) => {
								if (handleOriginal) {
									handleOriginal()
								}
							},
							style: {
								background: 'white',
								color: 'black'
						},
					}
					})}
			}

			columns.push(col);
		}

		return (
			<div id="main">
				<SelectInput/>
			 	<div id="tableCap"></div>
			  <ReactTable

				data={data}
				resolveData={data => data.map(row => row)}
				columns={columns}
				showPagination={false}
				minRows={0}
				sortable={false}
				resizable={false}
			  />
				<div className="bottom" id="option-container">
					<Link to="/">
						<button id="save" type="button">
							Save
						</button>
					</Link>
					<Link to="/availability" refresh="true">
						<button type="reset">
							Reset
						</button>
					</Link>
					<button onClick={this.showMenu}>
					  Additional Options
					</button>
					{
					  this.state.showMenu ? (
						  <div className="menu" ref={(element) => { this.state.dropdownMenu = element }}>
								<span id="blacklist">
									<div className="inputHeader">Professor Blacklist</div>
									<input id="profBlacklist" type="text" placeholder="Enter Professor Name"/>
									<button onClick={
										//possible check to make sure that user has indeed provided valid input.
											var profBL = document.getElementById("profBlacklist").value;
											var blacklistArray = [];
											function addProf (profBL) {
												blacklistArray.push(profBL);
												console.log("Blacklist: " + blacklistArray.join(", ") );
											}} id="profButton" form="main" type="profSave">
										Save
									</button>
								</span>
					  	</div>
						) : (null)
					}
				</div>
			</div>
		);
  }
}

//Time Select Input
export class SelectInput extends Component {
  render() {
    return (
      <div id="greetingText">Select your unavailability over a typical school week.</div>

    );
  }
};
