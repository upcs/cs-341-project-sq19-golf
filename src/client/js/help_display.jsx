import React, { Component } from 'react';
import {Link} from 'react-router-dom';
import '../css/styles.css';

//Options for About Page and Quit to Initial Page
export class HelpContainer extends Component {
	constructor(props) {
		super(props);

		let defaultOnClick = () => alert('UPSchedule is a convenient schedule planner created by students, for students.');
		this.state = {
			onClick: this.props.onClick || defaultOnClick
		};
	};

	render() {
		return (
		<section id="main">
			<div className="textContainer">
				<div className="contentsTable">
					<b id="tableTitle">Help Directory</b>
					<ul>
					<li><a href="#q1">Question 1</a></li>
					</ul>
				</div>
				<br />
				<div className="helpSet" id="h1">
					<a id="q1">Question</a>
					<p className="helpContent" id="a1">Answer</p>
				</div>

			</div>
		</section>
		);
	}
}
