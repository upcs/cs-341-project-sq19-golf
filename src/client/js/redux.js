import {createStore, combineReducers} from "redux";

//Describes all possible actions
export const actionTypes = {
  MODIFY_SCHEDULES: 'MODIFY_SCHEDULES',
  CLEAR_SCHEDULES: 'CLEAR_SCHEDULES',
  MODIFY_COURSE_CONTEXT: 'MODIFY_COURSE_CONTEXT',
  CLEAR_COURSE_CONTEXT: 'CLEAR_COURSE_CONTEXT',
  MODIFY_LAST_KEY: 'MODIFY_LAST_KEY',
  MODIFY_CONSTRAINTS: 'MODIFY_CONSTRAINTS'
};

//Initialize reducers
function scheduleReducer(viableSchedules = [], action) {
  switch (action.type) {
    case actionTypes.MODIFY_SCHEDULES:
      return action.viableSchedules;
    case actionTypes.CLEAR_SCHEDULES:
      return [];
    default:
      return viableSchedules;
  }
}

function courseReducer(courses = {}, action) {
  switch (action.type) {
    case actionTypes.MODIFY_COURSE_CONTEXT:
      return action.courses;
    case actionTypes.CLEAR_COURSE_CONTEXT:
      return {};
    default:
      return courses;
  }
}

function lastKeyReducer(lastKey = null, action) {
  switch (action.type) {
    case actionTypes.MODIFY_LAST_KEY:
      return action.lastKey;
    default:
      return lastKey;
  }
}

function scheduleConstraintReducer(constraints = null, action) {
  switch (action.type) {
    case actionTypes.MODIFY_CONSTRAINTS:
      return action.constraints;
    default:
      return constraints;
  }
}

export const rootReducer = combineReducers({
  viableSchedules: scheduleReducer,
  courseContext: courseReducer,
  lastKey: lastKeyReducer,
  constraints: scheduleConstraintReducer
});

//Initialize actions
export function modifySchedules(schedules) {
  return { type: actionTypes.MODIFY_SCHEDULES, viableSchedules: schedules };
}

export function modifyCourses(courses) {
  return { type: actionTypes.MODIFY_COURSE_CONTEXT, courses: courses };
}

export function modifyLastKey(lastKey) {
  return { type: actionTypes.MODIFY_LAST_KEY, lastKey: lastKey }
}

//TODO: Add professor blacklist
export function modifyConstraints(constraints) {
  return { type: actionTypes.MODIFY_CONSTRAINTS, constraints: constraints }
}

//Initialize store
export const store = createStore(rootReducer);
