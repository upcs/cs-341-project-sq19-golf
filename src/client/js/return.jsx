import React, {Component} from 'react';

export default class Return extends Component {
  constructor(props) {
    super(props);
  }

  render() {
    return (
      <div className="section">
        <button className="return" onClick={window.history.back()}>Return</button>
      </div>
    )
  }
}
