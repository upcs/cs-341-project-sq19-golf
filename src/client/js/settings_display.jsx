import React, { Component } from 'react';
import {Link} from 'react-router-dom';
import '../css/styles.css';

export class SettingsContainer extends Component {
  render() {
    return (
      <div>
      </div>
    )
  }
}
