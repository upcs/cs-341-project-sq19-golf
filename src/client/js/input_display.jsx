import React, {Component} from 'react';
import ReactDOM from 'react-dom';
import {Link} from 'react-router-dom';
import FadeIn from 'react-fade-in';
import fetch from 'node-fetch';
import $ from 'jquery';
import InputPredict from 'react-inline-predict';
import {store, modifySchedules, modifyLastKey} from './redux';
import Popup from 'reactjs-popup'
import '../css/styles.css';

//Menu bar
export class TopNavigation extends Component {
  constructor(props) {
		super(props);
	};

  backClick() {
    window.history.back();
  }

  render() {
    return (
      <section className="topNav">
        <div id="leftNav"><a>UP Scheduler</a></div>
        <div id='rightNav'>
          <Popup trigger={<button className="navEl"> Help </button>} modal closeOnDocumentClick>
            <div>
              <div className="header">Help<hr/></div>
              <span><ul align="left">
                <b>Availability Page</b><br/>
                Enter times you are unavailable, or professors you would like to avoid.
                Schedules matching these fields will not be generated
                <br/><br/>
                <b>Schedule Page</b><br/>
                View, name, and save your generated schedule
                </ul>
              </span>
            </div>
          </Popup>
          <Popup trigger={<button className="navEl"> About </button>} modal closeOnDocumentClick>
            <div>
              <div className="header">About<hr/></div>
              <ul align="left">UPSchedule is a convenient schedule planner created by students, for students.</ul>
            </div>
          </Popup>
        </div>
      </section>
    );
  }
};

//All input options
export class InputContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      totalInputs: 5,
      allCourses: {
        subjectMap: {},
        numberMap: {},
      },
      desiredCourses: []
    }

    this.handleCourseInputChange = this.handleCourseInputChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.retrieveAllCourses = this.retrieveAllCourses.bind(this);

    //Retrieve course data for auto-complete purposes
    this.retrieveAllCourses();
  }

  handleCourseInputChange(inputID, courseID, subject) {
    let desiredCourses = this.state.desiredCourses;
    desiredCourses[inputID] = {'subject': subject.toUpperCase(), 'courseID': courseID};
    this.setState({ 'desiredCourses': desiredCourses });
    this.modifyNecessaryInputs();
  }

  //Determine whether inputs should be added or removed, onChange()
  modifyNecessaryInputs() {
    let inputGroups = $("#classGroup").children();

    //Iterate over all inputs to determine fill status
    let populatedInputs = 0;
    for (let i = 0; i < inputGroups.length; i++) {
      let subjectInput = inputGroups[i].firstChild.children[0];
      let courseInput = inputGroups[i].lastChild.children[0];

      if (subjectInput.value !== '' && courseInput.value !== '') populatedInputs++;
    }

    //Add or remove inputs as necessary
    if (populatedInputs >= inputGroups.length) {
       this.setState({ 'totalInputs': this.state.totalInputs + 1 });
    }
    else if (populatedInputs < inputGroups.length - 1 && inputGroups.length > 5) {
       this.setState({ 'totalInputs': this.state.totalInputs - 1 });
    }
  }

  async retrieveAllCourses() {
    await fetch('/api/allCoursesRequest', {
      method: 'POST',
      body: '',
      headers: {
        'Accept': 'application/json',
        'Content-Type': 'application/json'
      }
    }).then(res => res.json()
    ).then(resJSON => {
      console.log(resJSON);

      this.setState({'allCourses': {'subjectMap': resJSON.subjMap, 'numberMap': resJSON.numMap} });
    }).catch((error) => {
      //console.log(error);
    });
  }

  async handleSubmit(event) {
    let desiredCourses = this.state.desiredCourses.filter(course => course.subject && course.courseID); //Remove undefined entries
    await fetch('/api/scheduleRequest', {
      method: 'POST',
      body: JSON.stringify(desiredCourses),
      headers: {
        'Accept': 'application/json',
        'Content-Type': 'application/json'
      }
    }).then(res => res.json()
    ).then(resJSON => {
      store.dispatch({ type: "CLEAR_SCHEDULES" }); //Hack to fix React's dumbass key-based rendering
      console.log(resJSON);
      store.dispatch(modifySchedules(resJSON));
    }).catch((error) => {
      //console.log(error);
    });
}

  render() {
    let courseList = [], references = {};
    for (let i = 0, idx = 0; i < this.state.totalInputs; i++, idx += 2) {
      let course = <CourseInput key={"course-" + i} id={i} idx={idx} value={this.state["course-" + i]}
                    courses={this.state.allCourses} references={references} onChange={this.handleCourseInputChange}/>
      courseList.push(course);
    }

    return (
      <div>
        <form id="main">
          <TermInput/>
          <div id="classGroup">
            {courseList}
          </div>
        </form>
        <div className="bottom">
          <Link to='/schedules'>
            <button form="main" onClick={this.handleSubmit}>
              Submit
            </button>
          </Link>
          <Link to='/availability'>
            <button form="main">
              Availability
            </button>
          </Link>
        </div>
      </div>
    );
  }
};

//Academic term input
export class TermInput extends Component {
  render() {
    return (
      <div id="greetingText">Select your courses for
        <input id="termSelect" type="text" defaultValue="Fall 2019"/>
      </div>
    );
  }
};

//Course selection input
export class CourseInput extends Component {
  constructor(props) {
    super(props);
    this.state ={
      input: {
        subject: null,
        number: null
      }
    }

    this.handleInput = this.handleInput.bind(this);
  }

  createRef(id) {
    if (this.props.references && !this.props.references.hasOwnProperty(id)) {
      return this.props.references[id] = React.createRef();
    }
    else if (this.props.references) {
      return this.props.references[id];
    }
  }

  //Focuses next input box
  _handleKeyDown(e, idx) {
    let lastKey = store.getState().lastKey;

    if (lastKey === "Shift" && e.key === "Tab") {
      let inputs = this.props.references;
      if (inputs.hasOwnProperty(idx - 1)) {
        ReactDOM.findDOMNode(inputs[idx - 1].current).children[0].focus();
      }
    }
    else if (e.key === "Tab") {
      let inputs = this.props.references;
      if (inputs.hasOwnProperty(idx + 1)) {
        ReactDOM.findDOMNode(inputs[idx + 1].current).children[0].focus();
      }
    }

    if (e.key === "Shift") store.dispatch(modifyLastKey(e.key));
  }

  _handleKeyUp(e) {
    if (e.key === "Shift") store.dispatch(modifyLastKey(null));
  }

  handleInput(value, context) {
    let input = this.state.input;
    if (context == 'subject') {
      input[context] = value.toString().toUpperCase();
    }
    else input[context] = value.toString();
    this.setState({ input });

    //Send input back up to parent component
    this.props.onChange(this.props.id, input.number, input.subject)
  }

  render() {
    let input = this.state.input;
    let courses = this.props.courses;

    console.log(input);
    console.log(courses);

    let numbers = (courses.subjectMap[input.subject]) ? courses.subjectMap[input.subject] : courses.subjectMap.all;
    let subjects = (courses.numberMap[input.number]) ? courses.numberMap[input.number] : courses.numberMap.all;

    return (
      <div className="classSelect">
        <InputPredict
          type="text"
          name="subject"
          placeholder="Course Subject"
          dictionary={subjects}
          ref={this.createRef(this.props.idx)}
          onKeyDown={(e) => this._handleKeyDown(e, this.props.idx)}
          onValueChange={((value) => this.handleInput(value, 'subject'))}
          onKeyUp={(e) => this._handleKeyUp(e)}
        />
        <InputPredict
          type="text"
          name="number"
          pattern="[0-9]*"
          placeholder="Course Number"
          dictionary={numbers}
          ref={this.createRef(this.props.idx + 1)}
          onValueChange={((value) => this.handleInput(value, 'number'))}
          onKeyDown={(e) => this._handleKeyDown(e, this.props.idx + 1)}
          onKeyUp={(e) => this._handleKeyUp(e)}
        />
      </div>
    );
  }
};
