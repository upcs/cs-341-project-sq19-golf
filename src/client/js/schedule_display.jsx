import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import $ from 'jquery';
import { saveAs } from 'browser-filesaver';
import * as html2canvas from 'html2canvas';
import Popup from 'reactjs-popup'
import {Link} from 'react-router-dom';
import {connect} from "react-redux";
import {store, modifySchedules} from './redux';
import '../css/styles.css';

//Displays all user-generated, viable schedules
export class SchedulesContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      scheduleName: 'schedules' //Default schedule name value for png generation
    }

    this.printDocument = this.printDocument.bind(this);
    this.handleScheduleName = this.handleScheduleName.bind(this);
  }

  //Redux-react store connection allowing for reactive state updates upon viableSchedules change
  connectSchedules() {
    const mapStateToProps = state => {
      return { viableSchedules: state.viableSchedules };
    };

    const schedulesList = ({ viableSchedules }) => {
      return viableSchedules.map((schedule, i) => {
        return <ScheduleDisplay key={"schedule-" + i} schedule={schedule}/>
      });
    };

    return connect(mapStateToProps)(schedulesList);
  }

  //Allows for the saving of the schedule display DOM section as a png
  printDocument() {
    const input = document.getElementById('divToPrint');
    if (input && input.childElementCount !== 0) {
      html2canvas(input).then((canvas) => {
        canvas.toBlob((blob) => {
          saveAs(blob, this.state.scheduleName + ".png");
        });
      });
    }
    else if (input) alert("Please generate a schedule first");
  }

  //Set png schedule name
  handleScheduleName(event) {
    this.setState({scheduleName: event.target.value});
  }

  render() {
    const Schedules = this.connectSchedules();

    return (
      <section id="main">
    		<div id="name">Schedule Name:
    			<input id="scheduleName" type="text" placeholder="Enter Schedule Name Here" onChange={this.handleScheduleName}/>
    		</div>
    		<div id="divToPrint" className="pdfdim">
  			  <Schedules/>
    		</div>
        <div className="bottom">
          <button onClick={this.printDocument}>Save As PNG</button>
          <button className="return" onClick={() => window.history.back()}>Return</button>
        </div>
      </section>
    );
  }
}

//Represents a single viable schedule
export class ScheduleDisplay extends Component {
  constructor(props) {
    super(props);
    this.state = {
      schedule: props.schedule || [],
      selected: false
    };

    this.updateDisplay = this.updateDisplay.bind(this);
  }

  //Handles display of schedule registration information onClick
  updateDisplay(truthVal) {
    this.setState({ selected: truthVal })
  }

  render() {
    //Render all scheduled classes
    let classDisplayList = this.state.schedule.map((classData, i) => {
      return (
        <ClassDisplay key={"class-" + i} classData={classData} updateDisplay={this.updateDisplay} selected={this.state.selected}/>
      );
    });

    //Attach and nest said classes within stylized schedule components
    return (
      <span className="scheduleOption">
    		<div className="scheduleContainer">
    			{classDisplayList}
    		</div>
      </span>
    )
  }
};

//Displays data for a single class
export class ClassDisplay extends Component {
  constructor(props) {
  	super(props);
  	this.state = { classData: props.classData || {} };
  }

  //TODO: Consider moving this up to the parent
  //Handles stateful ferrying of data from child to parent, concerning schedule registration display
  handleMouseClick() {
    let selected = this.props.selected;
	  this.props.updateDisplay(!selected);
  }

  render() {
    let classData = this.state.classData;
	  const tooltipStyle = {
		  display: this.props.selected ? 'block' : 'none'
	  }

    return (
		  <div className="scheduleClass">
  			<div onClick={this.handleMouseClick.bind(this)} className="classLabel">
  				{classData.subject} {classData.number}{classData.section}<br/>
  				{classData.days} {classData.start} - {classData.end}<br/>
  				{classData.location}
  			</div>
  			<div className="timeLabel" style={tooltipStyle}>
  				{classData.title}<br/>
  				{classData.professor}<br/>
  				CRN: {classData.crn}<br/>
  				Credits: {classData.credits}
  			</div>
		  </div>
    );
  }
};
