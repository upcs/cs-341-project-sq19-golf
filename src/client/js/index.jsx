import React from 'react';
import ReactDOM from 'react-dom';
import ReactTable from "react-table";

import {HashRouter, Route} from 'react-router-dom';
import FadeIn from 'react-fade-in';

import {TopNavigation, InputContainer} from './input_display';
import {AvailabilityContainer} from './availability_display';
import {SchedulesContainer} from './schedule_display';

const FadeInRoute = ({ component: Component, ...rest }) => (
  <Route {...rest} render={(props) => (
    <FadeIn>
      <Component {...props}/>
    </FadeIn>
  )}/>
);

const pageContainer = (
  <div id="pageContainer">
    <TopNavigation/>
    <FadeInRoute exact path="/" component={InputContainer}/>
    <FadeInRoute exact path="/availability" component={AvailabilityContainer}/>
    <FadeInRoute exact path="/schedules" component={SchedulesContainer}/>
  </div>
);

const inputPage = (
    <HashRouter>
      {pageContainer}
    </HashRouter>
);

ReactDOM.render(inputPage, document.getElementById('root'));
