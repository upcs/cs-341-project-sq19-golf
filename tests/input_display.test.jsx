import React from 'react';
import { shallow, render, setState, mount, find } from 'enzyme';
import {store} from '../src/client/js/redux';
import {TopNavigation, InputContainer, CourseInput, TermInput} from '../src/client/js/input_display.jsx';

describe('TopNavigation', () => {
  test('Should render correctly', () => {
    const topNav = shallow(<TopNavigation/>);
    expect(topNav).toMatchSnapshot();
  });

  test('Should prevent the user from re-visiting the previous page if no history exists',  () => {
    const topNav = shallow(<TopNavigation/>);
    window.history = null;

    expect(topNav.instance().backClick()).toBe(undefined);
  });

  test('Should react with modal popups', () => {
    const topNav = mount(<TopNavigation/>);

    //Test modal clicks
    topNav.find('#modal-1').simulate('click');
    topNav.find('#modal-2').simulate('click');
  });
});

describe('InputContainer', () => {
  test('Should render correctly', () => {
    const inputContainer = shallow(<InputContainer/>);
    expect(inputContainer).toMatchSnapshot();
    const event = Object.assign(jest.fn(), { preventDefault: () => {}});
    let courseData = {'inputID': 0, 'courseID': 341, 'subject': 'CS'};

    //Ensures necessary inputs exist
    //inputContainer.instance().modifyNecessaryInputs();
    //expect(inputContainer.instance().state.totalInputs).toEqual(5);

    //Handle course data dump
    inputContainer.instance().retrieveAllCourses();
    expect(inputContainer.instance().state.allCourses).toEqual({'subjectMap': {}, 'numberMap': {}});

    //Handle course input change
    inputContainer.instance().handleCourseInputChange(courseData.inputID, courseData.courseID, courseData.subject);
    expect(inputContainer.instance().state.desiredCourses).toEqual([{'courseID': 341, 'subject': 'CS'}]);

    //Handle submission
    inputContainer.instance().handleSubmit(event);
    expect(inputContainer.instance().props.handleSubmit).toBe();

    //Handle submission with no desired courses
    window.alert = () => {};
    inputContainer.instance().setState({ desiredCourses: [] })
    inputContainer.instance().handleSubmit(event);
    expect(inputContainer.instance().props.handleSubmit).toBe();

    //Handle submission with invalid courses
    inputContainer.instance().setState({ desiredCourses: [{ subject: null , courseID: null }]);
    inputContainer.instance().handleSubmit(event);
    expect(inputContainer.instance().props.handleSubmit).toBe();
  });
});

describe('CourseInput', () => {
  let key = {'key': null};
  let courses = {'subjectMap': {'all': null}, 'numberMap': {'all': null}};

  test('Should render correctly', () => {
    const courseInput = shallow(<CourseInput courses={courses} />);
    expect(courseInput).toMatchSnapshot();
  });

  test('Should properly create references', () => {
    const courseInput = shallow(<CourseInput courses={courses}/>);

    courseInput.instance().createRef(0);
    expect(courseInput.instance().props.createRef).toEqual(undefined); //TODO: Fix
  });

  test('Should handle a TAB key press correctly', () => {
    const courseInput = shallow(<CourseInput courses={courses} references={{}} lastKey={key}/>);

    courseInput.instance()._handleKeyDown({key: "Tab"}, 0);
    expect(courseInput.instance().props.lastKey['key']).toEqual(null); //TODO: Fix
  });

  test('Should handle a SHIFT key press correctly', () => {
    const courseInput = shallow(<CourseInput courses={courses} references={{}} lastKey={key}/>);

    courseInput.instance()._handleKeyDown({key: "Shift"}, 0);
    expect(courseInput.instance().props.lastKey['key']).toEqual(null); //TODO: Fix
  });

  test('Should handle an arbitrary key press correctly', () => {
    const courseInput = shallow(<CourseInput courses={courses} lastKey={key}/>);

    courseInput.instance()._handleKeyUp({key: "Enter"}, 0);
    expect(courseInput.instance().props.lastKey['key']).toEqual(null); //TODO: Fix
  });

  test('Should write input to parent\'s state', () => {
    const courseInput = shallow(<CourseInput courses={courses} onChange={() => {}}/>);

    courseInput.instance().handleInput('test', false, 'subject');
    expect(courseInput.instance().state.input.subject).toEqual('TEST');
  });
});

describe('TermInput', () => {
  test('Should render correctly', () => {
    const termInput = shallow(<TermInput/>);
    expect(termInput).toMatchSnapshot();
  });
});
