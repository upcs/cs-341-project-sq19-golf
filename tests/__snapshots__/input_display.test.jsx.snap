// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseInput Should render correctly 1`] = `
<div
  className="classSelect"
>
  <InputPredict
    autocomplete="off"
    dictionary={null}
    name="subject"
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onValueChange={[Function]}
    placeholder="Course Subject"
    type="text"
  />
  <InputPredict
    autocomplete="off"
    dictionary={null}
    name="number"
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onValueChange={[Function]}
    pattern="[0-9]*"
    placeholder="Course Number"
    type="text"
  />
</div>
`;

exports[`InputContainer Should render correctly 1`] = `
<div>
  <form
    id="main"
  >
    <TermInput />
    <div
      id="classGroup"
    >
      <CourseInput
        courses={
          Object {
            "numberMap": Object {},
            "subjectMap": Object {},
          }
        }
        id={0}
        idx={0}
        key="course-0"
        onChange={[Function]}
        references={Object {}}
      />
      <CourseInput
        courses={
          Object {
            "numberMap": Object {},
            "subjectMap": Object {},
          }
        }
        id={1}
        idx={2}
        key="course-1"
        onChange={[Function]}
        references={Object {}}
      />
      <CourseInput
        courses={
          Object {
            "numberMap": Object {},
            "subjectMap": Object {},
          }
        }
        id={2}
        idx={4}
        key="course-2"
        onChange={[Function]}
        references={Object {}}
      />
      <CourseInput
        courses={
          Object {
            "numberMap": Object {},
            "subjectMap": Object {},
          }
        }
        id={3}
        idx={6}
        key="course-3"
        onChange={[Function]}
        references={Object {}}
      />
      <CourseInput
        courses={
          Object {
            "numberMap": Object {},
            "subjectMap": Object {},
          }
        }
        id={4}
        idx={8}
        key="course-4"
        onChange={[Function]}
        references={Object {}}
      />
    </div>
  </form>
  <div
    className="bottom"
  >
    <Link
      replace={false}
      to="/schedules"
    >
      <button
        form="main"
        onClick={[Function]}
      >
        Submit
      </button>
    </Link>
    <Link
      replace={false}
      to="/availability"
    >
      <button
        form="main"
      >
        Availability
      </button>
    </Link>
  </div>
</div>
`;

exports[`TermInput Should render correctly 1`] = `
<div
  id="greetingText"
>
  Select your courses for
  <input
    defaultValue="Fall 2019"
    disabled={true}
    id="termSelect"
    type="text"
  />
</div>
`;

exports[`TopNavigation Should render correctly 1`] = `
<section
  className="topNav"
>
  <div
    id="leftNav"
  >
    <a>
      UP Scheduler
    </a>
  </div>
  <div
    id="rightNav"
  >
    <Popup
      arrow={true}
      arrowStyle={Object {}}
      className=""
      closeOnDocumentClick={true}
      closeOnEscape={true}
      contentStyle={Object {}}
      defaultOpen={false}
      disabled={false}
      keepTooltipInside={false}
      lockScroll={false}
      modal={true}
      mouseEnterDelay={100}
      mouseLeaveDelay={100}
      offsetX={0}
      offsetY={0}
      on={
        Array [
          "click",
        ]
      }
      onClose={[Function]}
      onOpen={[Function]}
      open={false}
      overlayStyle={Object {}}
      position="bottom center"
      repositionOnResize={true}
      trigger={
        <button
          className="navEl"
          id="modal-1"
        >
           Help 
        </button>
      }
    >
      <div>
        <div
          className="header"
        >
          Help
          <hr />
        </div>
        <span>
          <ul
            align="left"
          >
            <b>
              Availability Page
            </b>
            <br />
            Enter times you are unavailable, or professors you would like to avoid. Schedules matching these fields will not be generated
            <br />
            <br />
            <b>
              Schedule Page
            </b>
            <br />
            View, name, and save your generated schedule
          </ul>
        </span>
      </div>
    </Popup>
    <Popup
      arrow={true}
      arrowStyle={Object {}}
      className=""
      closeOnDocumentClick={true}
      closeOnEscape={true}
      contentStyle={Object {}}
      defaultOpen={false}
      disabled={false}
      keepTooltipInside={false}
      lockScroll={false}
      modal={true}
      mouseEnterDelay={100}
      mouseLeaveDelay={100}
      offsetX={0}
      offsetY={0}
      on={
        Array [
          "click",
        ]
      }
      onClose={[Function]}
      onOpen={[Function]}
      open={false}
      overlayStyle={Object {}}
      position="bottom center"
      repositionOnResize={true}
      trigger={
        <button
          className="navEl"
          id="modal-2"
        >
           About 
        </button>
      }
    >
      <div>
        <div
          className="header"
        >
          About
          <hr />
        </div>
        <ul
          align="left"
        >
          UPSchedule is a convenient schedule planner created by students, for students.
        </ul>
      </div>
    </Popup>
  </div>
</section>
`;
